pkg/controller.v1/tensorflow/pod.go

```diff
@@ -292,19 +108,92 @@ func isDistributed(tfjob *tfv1.TFJob) bool {
 	return distributionCount != 1
 }
 
-func setRestartPolicy(podTemplateSpec *v1.PodTemplateSpec, spec *common.ReplicaSpec) {
-	if spec.RestartPolicy == common.RestartPolicyExitCode {
+func setRestartPolicy(podTemplateSpec *v1.PodTemplateSpec, spec *commonv1.ReplicaSpec) {
+	// This is necessary since restartPolicyExitCode is not supported in v1.PodTemplateSpec
+	if spec.RestartPolicy == commonv1.RestartPolicyExitCode {
 		podTemplateSpec.Spec.RestartPolicy = v1.RestartPolicyNever
 	} else {
 		podTemplateSpec.Spec.RestartPolicy = v1.RestartPolicy(spec.RestartPolicy)
 	}
 }
 
-func (tc *TFController) isNonGangSchedulerSet(tfjob *tfv1.TFJob) bool {
-	for _, spec := range tfjob.Spec.TFReplicaSpecs {
-		if spec.Template.Spec.SchedulerName != "" && spec.Template.Spec.SchedulerName != tc.Config.GangSchedulerName {
-			return true
+// IsWorker0Completed return true if pod of worker0 succeeded and exited with 0
+func (tc *TFController) IsWorker0Completed(tfjob *tfv1.TFJob, replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec) (bool, error) {
+	worker0Completed := false
+	logger := tflogger.LoggerForReplica(tfjob, strings.ToLower(string(tfv1.TFReplicaTypeWorker)))
+
+	pods, err := tc.GetPodsForJob(tfjob)
+	if err != nil {
+		tflogger.LoggerForJob(tfjob).Warnf("getPodsForTFJob error %v", err)
+		return false, err
+	}
+
+	// Get all pods for the type rt.
+	pods, err = tc.FilterPodsForReplicaType(pods, strings.ToLower(string(tfv1.TFReplicaTypeWorker)))
+	if err != nil {
+		return false, err
+	}
+
+	podSlices := tc.GetPodSlices(pods, int(*replicas[tfv1.TFReplicaTypeWorker].Replicas), logger)
+	for index, podSlice := range podSlices {
+		if len(podSlice) == 1 {
+			pod := podSlice[0]
+			// Get the exit code of the tensorflow container.
+			var exitCode int32 = 0xbeef // magic number
+			for _, status := range pod.Status.ContainerStatuses {
+				state := status.State
+				if status.Name == tfv1.DefaultContainerName && state.Terminated != nil {
+					exitCode = state.Terminated.ExitCode
+				}
+			}
+
+			if index == 0 && exitCode == 0 && pod.Status.Phase == v1.PodSucceeded {
+				worker0Completed = true
+			}
+		}
+	}
+	return worker0Completed, nil
+}
+
+// PodRestart return true if pod failed with retryable exit code
+func (tc *TFController) PodRestart(tfjob *tfv1.TFJob, spec *commonv1.ReplicaSpec, rtype commonv1.ReplicaType) (bool, error) {
+	podRestart := false
+	logger := tflogger.LoggerForReplica(tfjob, strings.ToLower(string(rtype)))
+
+	pods, err := tc.GetPodsForJob(tfjob)
+	if err != nil {
+		tflogger.LoggerForJob(tfjob).Warnf("getPodsForTFJob error %v", err)
+		return false, err
+	}
+
+	// Get all pods for the type rt.
+	pods, err = tc.FilterPodsForReplicaType(pods, strings.ToLower(string(rtype)))
+	if err != nil {
+		return false, err
+	}
+
+	podSlices := tc.GetPodSlices(pods, int(*spec.Replicas), logger)
```

Comment:
minor: the logic here to get list of pods and extract pod status is very close to `isWorker0Complete`. Probably can do some refactor to reuse same code snippet.